pkg/rtc/mediatracksubscriptions.go

```diff
@@ -0,0 +1,499 @@
+	_, ok := t.subscribedTracks.Load(subID)
+	return ok
+}
+
+// AddSubscriber subscribes sub to current mediaTrack
+func (t *MediaTrackSubscriptions) AddSubscriber(sub types.Participant, codec webrtc.RTPCodecCapability, wr WrappedReceiver) (*sfu.DownTrack, error) {
+	subscriberID := sub.ID()
+
+	// don't subscribe to the same track multiple times
+	if _, ok := t.subscribedTracks.Load(subscriberID); ok {
+		return nil, nil
+	}
+
+	var rtcpFeedback []webrtc.RTCPFeedback
+	switch t.params.MediaTrack.Kind() {
+	case livekit.TrackType_AUDIO:
+		rtcpFeedback = t.params.SubscriberConfig.RTCPFeedback.Audio
+	case livekit.TrackType_VIDEO:
+		rtcpFeedback = t.params.SubscriberConfig.RTCPFeedback.Video
+	}
+	downTrack, err := sfu.NewDownTrack(webrtc.RTPCodecCapability{
+		MimeType:     codec.MimeType,
+		ClockRate:    codec.ClockRate,
+		Channels:     codec.Channels,
+		SDPFmtpLine:  codec.SDPFmtpLine,
+		RTCPFeedback: rtcpFeedback,
+	}, wr, t.params.BufferFactory, subscriberID, t.params.ReceiverConfig.PacketBufferSize)
+	if err != nil {
+		return nil, err
+	}
+	subTrack := NewSubscribedTrack(SubscribedTrackParams{
+		PublisherID:       t.params.MediaTrack.ParticipantID(),
+		PublisherIdentity: t.params.MediaTrack.ParticipantIdentity(),
+		SubscriberID:      subscriberID,
+		MediaTrack:        t.params.MediaTrack,
+		DownTrack:         downTrack,
+	})
+
+	var transceiver *webrtc.RTPTransceiver
+	var sender *webrtc.RTPSender
+	if sub.ProtocolVersion().SupportsTransceiverReuse() {
+		//
+		// AddTrack will create a new transceiver or re-use an unused one
+		// if the attributes match. This prevents SDP from bloating
+		// because of dormant transceivers building up.
+		//
+		sender, err = sub.SubscriberPC().AddTrack(downTrack)
+		if err != nil {
+			return nil, err
+		}
+
+		// as there is no way to get transceiver from sender, search
+		for _, tr := range sub.SubscriberPC().GetTransceivers() {
+			if tr.Sender() == sender {
+				transceiver = tr
+				break
+			}
+		}
+		if transceiver == nil {
+			// cannot add, no transceiver
+			return nil, errors.New("cannot subscribe without a transceiver in place")
+		}
+	} else {
+		transceiver, err = sub.SubscriberPC().AddTransceiverFromTrack(downTrack, webrtc.RTPTransceiverInit{
+			Direction: webrtc.RTPTransceiverDirectionSendonly,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		sender = transceiver.Sender()
+		if sender == nil {
+			// cannot add, no sender
+			return nil, errors.New("cannot subscribe without a sender in place")
+		}
+	}
+
+	sendParameters := sender.GetParameters()
+	downTrack.SetRTPHeaderExtensions(sendParameters.HeaderExtensions)
+
+	downTrack.SetTransceiver(transceiver)
+	// when outtrack is bound, start loop to send reports
+	downTrack.OnBind(func() {
+		go subTrack.Bound()
+		go t.sendDownTrackBindingReports(sub)
+	})
+	downTrack.OnPacketSent(func(_ *sfu.DownTrack, size int) {
+		if t.params.Telemetry != nil {
```

Comment:
Add the not nil checks in each of these as cloud is not setting this service. I am assuming it would be needed in cloud also. We can remove the ifs here when this is set in cloud also.

Also, maybe set trackID outside and use it in the closure. Calling a function on every packet to get trackID seems like something that should be avoided. Also, wondering if a bunch of these can actually be moved into sfu.DownTrack and pull information periodically.

Reply:
that's a good suggestion. It does seem that if avoidable, we should avoid per-packet operations. CC @ashellunts @frostbyte73